core/src/main/java/com/sequenceiq/cloudbreak/controller/StackCreatorService.java

```diff
@@ -379,9 +386,68 @@ void fillInstanceMetadata(DetailedEnvironmentResponse environment, Stack stack)
                 instanceMetaData.setPrivateId(privateIdNumber++);
                 instanceMetaData.setInstanceStatus(InstanceStatus.REQUESTED);
             }
-            multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroup);
-            prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
         }
+        if (StackType.DATALAKE.equals(stack.getType())
+                && containsIgnoreCase(stack.getCluster().getBlueprint().getBlueprintText(), SDX_ENTERPRISE_DATALAKE_TEXT)) {
+            datalakeMultiAZCalculator(stack, subnetAzPairs, stackSubnetId, stackAz);
+        } else {
+            for (InstanceGroup instanceGroup : sortInstanceGroups(stack)) {
+                multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroup.getInstanceGroupNetwork(),
+                        instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet());
+                prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
+            }
+        }
+    }
+
+    public void datalakeMultiAZCalculator(Stack stack, Map<String, String> subnetAzPairs, String stackSubnetId, String stackAz) {
```

Comment:
From the method name and the size of the logic it seems that could be moved to a separate class or move this logic into the `MultiAzCalculatorService`

Reply:
or better to say create a DataLakeMultiAzCalculatorService which extends MultiAzCalculatorService

Reply:
addressed.